### PR TITLE
Changing peer selection process & show connection errors in debug page

### PR DESCRIPTION
**Pull request by mm-near**
Idea on how we could do a small change to peer selection process, by preferring to pick peers that we've used to connect to, in order to speed up connection process.

And also updated network debug page to show the reason why the connection to peer has failed.

### CODE REVIEW

**Comment by mm-near** [comment]
@pompon0 - PTAL and let me know what you think. If this approach makes sense, I'll add tests and we can ship it.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 1]
```diff
@@ -869,6 +876,17 @@ impl PeerManagerActor {
...
+            error!(target: "network", ?err, "Failed to update peers last seen time.");
+        }
+
+        // Get the peers that we failed to connect to recently, and mark their connection as failed.
+        for failed_peer_id in self.failed_connections.lock().expect("Lock error").drain(..) {
+            debug!(target: "network", ?failed_peer_id, "Marking peer as failed.");
+            if self.peer_store.peer_connection_failed(&self.clock, &failed_peer_id).is_err() {
+                error!(target: "network", ?failed_peer_id, "Failed to mark peer as failed.");
+            }
+        }
```
Comment (pompon0) [inline chain/network/src/peer_manager/peer_manager_actor.rs:889]: this should happen synchronously in unregister() call.

Comment (pompon0) [reply chain/network/src/peer_manager/peer_manager_actor.rs:889]: or, if you want to track TCP handshake failure, synchronously below (in the place you do the push_back to the queue). There is no point in caching it.

File: chain/network/src/peer_manager/peer_store/mod.rs [hunk 2]
```diff
@@ -266,6 +313,26 @@ impl PeerStore {
...
         ignore_fn: impl Fn(&KnownPeerState) -> bool,
     ) -> Option<PeerInfo> {
+        // With some odds - try picking one of the 'NotConnected' peers -- these are the ones that we were able to connect to in the past.
+        if thread_rng().gen_bool(PICK_KNOWN_PEER_ODDS) {
+            let preferred_peer = self.find_peers(
+                |p| {
+                    (p.status == KnownPeerStatus::NotConnected)
+                        && !ignore_fn(p)
+                        && p.peer_info.addr.is_some()
+                        // if we're connecting only to the bood nodes - filter out the nodes that are not bootnodes.
```
Comment (pompon0) [inline chain/network/src/peer_manager/peer_store/mod.rs:323]: boot

File: chain/epoch-manager/src/proposals.rs [hunk 3]
```diff
@@ -270,15 +269,51 @@ mod old_validator_selection {
...
+        let zone = u64::MAX - (u64::MAX - bound + 1) % (bound);
+        loop {
+            let v = rng.next_u64();
+            let mul = (v as u128) * (bound as u128);
+            let (hi, lo) = ((mul >> 64) as u64, mul as u64);
+            if lo < zone {
+                return hi;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mm-near) [reply chain/epoch-manager/src/proposals.rs]: fixed.

File: chain/network/src/peer_manager/peer_manager_actor.rs (hunk 1 above)
Comment (mm-near) [reply chain/network/src/peer_manager/peer_manager_actor.rs]: done.

File: chain/network/src/peer_manager/peer_store/mod.rs (hunk 2 above)
Comment (mm-near) [reply chain/network/src/peer_manager/peer_store/mod.rs]: done.

File: chain/network/src/peer_manager/peer_store/mod.rs [hunk 4]
```diff
@@ -378,6 +412,31 @@ impl PeerStore {
         Ok(())
     }
 
+    /// Records the last attempt to connect to peer.
+    /// Marks the peer as Uknown (as we failed to connect to it).
```
Comment (pompon0) [inline chain/network/src/peer_manager/peer_store/mod.rs:416]: Unknown

File: chain/network/src/types.rs [hunk 5]
```diff
@@ -95,6 +101,9 @@ pub struct KnownPeerState {
     pub status: KnownPeerStatus,
     pub first_seen: time::Utc,
     pub last_seen: time::Utc,
+    // Last time we tried to connect to this peer.
+    // This data is not persisted in storage.
```
Comment (pompon0) [inline chain/network/src/types.rs:105]: if you don't want to persist this field, then Option<(time::Utc, anyhow::Result)> would be better.

**Pull request merged**